sklearn/utils/stats.py

```diff
@@ -25,37 +26,99 @@ def _weighted_percentile(array, sample_weight, percentile=50):
         Percentile to compute. Must be value between 0 and 100.
 
+    interpolation : {"linear", "lower", "higher"}, default="linear"
+        This optional parameter specifies the interpolation method to
+        use when the desired percentile lies between two data points
+        ``i < j``:
+        * 'linear': ``i + (j - i) * fraction``, where ``fraction``
+          is the fractional part of the index surrounded by ``i``
+          and ``j``.
+        * 'lower': ``i``.
+        * 'higher': ``j``.
+        * 'nearest': ``i`` or ``j``, whichever is nearest.
+
     Returns
     -------
-    percentile : int if `array` 1D, ndarray if `array` 2D
+    percentile_value : int if `array` 1D, ndarray if `array` 2D
         Weighted percentile.
     """
+    possible_interpolation = ("linear", "lower", "higher", "nearest")
+    if interpolation not in possible_interpolation:
+        raise ValueError(
+            f"'interpolation' should be one of "
+            f"{', '.join(possible_interpolation)}. Got '{interpolation}' "
+            f"instead."
+        )
+
     n_dim = array.ndim
     if n_dim == 0:
         return array[()]
     if array.ndim == 1:
         array = array.reshape((-1, 1))
-    # When sample_weight 1D, repeat for each array.shape[1]
+
     if (array.shape != sample_weight.shape and
             array.shape[0] == sample_weight.shape[0]):
+        # when `sample_weight` is 1D, we repeat it for each column of `array`
         sample_weight = np.tile(sample_weight, (array.shape[1], 1)).T
+
+    n_rows, n_cols = array.shape
+
     sorted_idx = np.argsort(array, axis=0)
     sorted_weights = _take_along_axis(sample_weight, sorted_idx, axis=0)
-
-    # Find index of median prediction for each sample
+    percentile = np.array([percentile / 100] * n_cols)
     weight_cdf = stable_cumsum(sorted_weights, axis=0)
-    adjusted_percentile = percentile / 100 * weight_cdf[-1]
-    percentile_idx = np.array([
-        np.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
-        for i in range(weight_cdf.shape[1])
-    ])
-    percentile_idx = np.array(percentile_idx)
-    # In rare cases, percentile_idx equals to sorted_idx.shape[0]
-    max_idx = sorted_idx.shape[0] - 1
-    percentile_idx = np.apply_along_axis(lambda x: np.clip(x, 0, max_idx),
-                                         axis=0, arr=percentile_idx)
-
-    col_index = np.arange(array.shape[1])
-    percentile_in_sorted = sorted_idx[percentile_idx, col_index]
-    percentile = array[percentile_in_sorted, col_index]
-    return percentile[0] if n_dim == 1 else percentile
+
+    def _squeeze_arr(arr, n_dim):
+        return arr[0] if n_dim == 1 else arr
+
+    if interpolation == "nearest":
+        # compute by nearest-rank method
+        adjusted_percentile = percentile * weight_cdf[-1]
+        percentile_value_idx = np.array([
+            np.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
+            for i in range(weight_cdf.shape[1])
+        ])
```

Comment:
I am not sure if this gives you the index of the nearest element.
```
a = np.array([0.2, 0.49, 0.6, 0.8, 1])
np.searchsorted(a, 0.5, side='left')
```
> 2

If anything I think it gives you `j` (for ``i < j``). It returns the index that satisfies: `a[i-1] < v <= a[i]`. i.e. your desired value (`0.5`) is bigger than the element on the left `i` or 0.49 and smaller than the element on the right `j` or 0.6

Reply:
The index it gives is where your desired number 'should go' to keep order. That index is the index of the bigger element (for two data points `i < j`)

Reply:
(nevermind, we worked out wiki definition of nearest = numpy 'higher')